proposals/168/making_spark_visitors_extensible.md

```diff
@@ -0,0 +1,271 @@
+    Collection<CustomFacetBuilder<Object, OpenLineage.JobFacet>> jobFacetBuilders;
+    SparkContext sparkContext;
+}
+```
+
+The `OpenLineageEventListenerFactory#setOpenLineageContext` method will be called once the concrete service 
+implementation has been constructed, before the first `createX` method is called.
+
+Each of the `createX` methods will be called exactly once during a Spark application. There's no guarantee regarding the
+order in which the `create` methods will be called. Thus, the constructed builders should make no assumptions about the
+state of the `OpenLineageContext` when the `create` method is called. E.g., the `inputDatasetFacetBuilders` should not 
+assume that all of the `inputDatasetQueryPlanVisitors` have been instantiated or that the `Collection` returned by the
+getter has been fully populated. Builders that need to depend on the ability to delegate to other builders should store
+a reference to the `OpenLineageContext` and call the appropriate getter at the time the delegate is actually needed.
+
+The builders returned are all some form of `PartialFunction` - they must all define an `isDefinedAt(T event)` method
+which determines the type of argument the builder can respond to. The `createInputDatasetBuilder` and 
+`createOutputDatasetBuilder` methods return `PartialFunction`s that construct either an `InputDatasetBuilder` or an
+`OutputDatasetBuilder`, respectively. Builders are returned so that facets constructed by the other `PartialFunction`s
+can be attached to the dataset before the immutable `InputDataset` or `OutputDataset` is returned. It is expected that
+the `CustomFacet`s returned are already immutable.
+
+For each `SparkListener` event, an `OpenLineage.RunEvent` will be constructed by 
+1. Construct `RunFacet`s by invoking each of the `runFacetBuilders`
+2. Construct `JobFacet`s by invoking each of the `jobFacetBuilders`
+3. Construct `InputDatasetBuilder`s by invoking each of the `inputDatasetBuilder`s.
+4. _If_ any `InputDatasetBuilder`s are returned, invoke the `inputDatasetFacetBuilder`s and attach any returned facets to
+   the `InputDatasetBuilder`s
+5. _If_ any `InputDatasetBuilder`s are returned, invoke the `datasetFacetBuilder`s and attach any returned facets to 
+    the `InputDatasetBuilder`s
+6. Construct `OutputDatasetBuilder`s by invoking each of the `outputDatasetBuilder`s.
+7. _If_ any `OuptutDatasetBuilder`s are returned, invoke the `outputDatasetFacetBuilder`s and attach any returned facets to
+   the `OutputDatasetBuilder`s
+5. _If_ any `OutputDatasetBuilder`s are returned, invoke the `datasetFacetBuilder`s and attach any returned facets to
+   the `OutputDatasetBuilder`s
+
+`InputDatasetFacet`s and `OutputDatasetFacet`s will be attached to _any_ `InputDatasetBuilder` or `OutputDatasetBuilder`
+found for the event. This is because facets may be constructed from generic information that is not specifi**cally tied
```

Comment:
> Does that mean that the found facets will be attached also to datasets constructed outside that particular factory implementation?

Yes - the CustomFacetBuilder may return a very generic facet, such as `OutputStatistics`, which is applicable to any `OutputDataset`. I see two approaches:

1. The Dataset builders must be responsible for building the Datasets and the facets. This means the Dataset builder must always match on the highest level event - e.g., the Stage or the ActiveJob - and then descend down to find the RDD or whatever is needed to construct the Dataset and all of the associated facets. This also means that only facets defined by the Dataset builder can be attached to that Dataset.
2. Each builder is responsible for constructing exactly one component - Dataset builders only construct Datasets and Facet builders only construct Facets. This means each builder can be built specific to the event that actually contains the data necessary to construct the Dataset/Facet (i.e., the Stage or the RDD) while the logic of moving from high-level event to low-level event can be extracted out and that facets can be attached to any dataset regardless of whether the Dataset builder is aware of that facet. 

The approach outlined is the latter, as I feel it is more extensible and allows authors to attach facets to Datasets that are perhaps custom to the author (e.g., there might be some business metadata I want to attach to the specific BigQuery Dataset that I'm processing even though the BigQuery library author doesn't know anything about my business metadata). From my experimentation, there is generally only one input Dataset attached to a Stage, so the facets, I think, can reliably be attached to the correct input dataset. The OutputDataset must typically be found in the QueryExecution and there's only one output dataset per QueryExecution, so any output dataset facets can be reliably attached to that. 

Let me know if this reasoning makes sense to you.